connectors/sources/sharepoint_online.py

```diff
@@ -626,20 +626,82 @@ async def site_users(self, site_web_url):
         except NotFound:
             return
 
-    async def sites(self, parent_site_id, allowed_root_sites):
-        select = ""
+    async def sites(
+        self,
+        sharepoint_host,
+        allowed_root_sites,
+        enumerate_all_sites=True,
+        fetch_subsites=False,
+    ):
+        if allowed_root_sites == [WILDCARD] or enumerate_all_sites:
+            self._logger.debug(f"Looking up all sites to fetch: {allowed_root_sites}")
+            async for site in self._all_sites(sharepoint_host, allowed_root_sites):
+                yield site
+        else:
+            self._logger.debug(f"Looking up individual sites: {allowed_root_sites}")
+            for allowed_site in allowed_root_sites:
+                try:
+                    if fetch_subsites:
+                        async for site in self._fetch_site_and_subsites_by_path(
+                            sharepoint_host, allowed_site
+                        ):
+                            yield site
+                    else:
+                        yield await self._fetch_site(sharepoint_host, allowed_site)
+
+                except NotFound:
+                    self._logger.warning(
+                        f"Could not look up site '{allowed_site}' by relative path in parent site: {sharepoint_host}"
+                    )
 
+    async def _all_sites(self, sharepoint_host, allowed_root_sites):
+        select = ""
         async for page in self._graph_api_client.scroll(
-            f"{GRAPH_API_URL}/sites/{parent_site_id}/sites?search=*&$select={select}"
+            f"{GRAPH_API_URL}/sites/{sharepoint_host}/sites?search=*&$select={select}"
         ):
             for site in page:
                 # Filter out site collections that are not needed
-                if (
-                    WILDCARD not in allowed_root_sites
-                    and site["name"] not in allowed_root_sites
-                ):
+                if [WILDCARD] != allowed_root_sites and site[
```

Comment:
following on Chenhui comment: should we check `if WILDCARD not in allowed_root_sites ...`